scripts/run_changed.js

```diff
@@ -52,35 +52,82 @@ const alwaysRunTestPaths = [
   'integration/webpack'
 ];
 
+/**
+ * These files trigger tests in other dirs
+ */
+const specialPaths = {
+  'scripts/emulator-testing/emulators/firestore-emulator.ts': ['packages/firestore'],
+  'scripts/emulator-testing/emulators/database-emulator.ts': ['packages/database'],
+  'scripts/emulator-testing/emulators/emulator.ts': ['packages/firestore', 'packages/database'],
+  'scripts/emulator-testing/firestore-test-runner.ts': ['packages/firestore'],
+  'scripts/emulator-testing/database-test-runner.ts': ['packages/database']
+};
+
 /**
  * Identify modified packages that require tests.
  */
 async function getChangedPackages() {
+  const packageInfo = JSON.parse(
+    (await exec('npx lerna ls --json', { cwd: root })).stdout
+  );
+  const depGraph = JSON.parse(
+    (await exec('npx lerna ls --graph', { cwd: root })).stdout
+  );
   const diff = await git.diff(['--name-only', 'origin/master...HEAD']);
   const changedFiles = diff.split('\n');
   const changedPackages = {};
   for (const filename of changedFiles) {
+    // Files that trigger full test suite.
     if (fullTestTriggerFiles.includes(filename)) {
       console.log(
         chalk`{blue Running all tests because ${filename} was modified.}`
       );
       return { testAll: true };
     }
+    // Files outside a package dir that should trigger its tests.
+    if (specialPaths[filename]) {
+      for (const targetPackage of specialPaths[filename]) {
+        changedPackages[targetPackage] = 'dependency';
+      }
+    }
+    // Check for changed files inside package dirs.
     const match = filename.match('^(packages/[a-zA-Z0-9-]+)/.*');
     if (match && match[1]) {
-      const pkg = require(resolve(root, match[1], 'package.json'));
-      if (pkg && pkg.scripts.test) {
-        changedPackages[match[1]] = true;
+      const changedPackage = require(resolve(root, match[1], 'package.json'));
+      if (changedPackage && changedPackage.scripts.test) {
+        // Add the package itself.
+        changedPackages[match[1]] = 'direct';
+        // Add packages that depend on it.
+        for (const package in depGraph) {
```

Comment:
Shouldn't we run tests for depending packages even if `changedPackage.scripts.test` evaluates to false? something like:
```
if (changedPackage) {
  if (changedPackage.scripts.test) {
      changedPackages[match[1]] = 'direct';
  }

  for (const package in depGraph) {
  ...
  }
}
```

Reply:
Yes, fixed.